mino/minogrpc/routing/routing.go

```diff
@@ -0,0 +1,428 @@
+	"go.dedis.ch/fabric/mino"
+	"golang.org/x/xerrors"
+)
+
+//go:generate protoc -I ./ --go_out=plugins=grpc:./ ./message.proto
+
+var eachLine = regexp.MustCompile(`(?m)^(.+)$`)
+
+// Factory defines the primitive to create a Routing
+type Factory interface {
+	FromIterator(mino.AddressIterator) (Routing, error)
+	FromAny(*any.Any) (Routing, error)
+}
+
+// Routing defines the functions needed to route messages
+type Routing interface {
+	encoding.Packable
+	// GetRoute should return the gateway address for a corresponding addresse.
+	// In a tree communication it is typically the address of the child that
+	// contains the "to" address in its sub-tree.
+	GetRoute(from, to mino.Address) (gateway mino.Address, err error)
+	// GetDirectLinks return the direct links of the elements. In a tree routing
+	// this is typically the children of the node.
+	GetDirectLinks(from mino.Address) ([]mino.Address, error)
+}
+
+// TreeRouting holds the routing tree of a network. It allows each node of the
+// tree to know which child it should contact in order to relay a message that
+// is in it sub-tree.
+//
+// - implements Routing
+type TreeRouting struct {
+	root         *treeNode
+	routingNodes map[string]*treeNode
+}
+
+// NewTreeRouting ...
+func NewTreeRouting(height int) *TreeRouting {
+	return &TreeRouting{}
+}
+
+// TreeRoutingOpts is the implementation of treeTreeRoutingOpts
+var TreeRoutingOpts = treeRoutingOpts{
+	// Addr is the address of the node, the value should be of type mino.Address
+	Addr: "addr",
+	// TreeHeight is the maximum tree height
+	TreeHeight: "treeHeight",
+}
+
+type treeRoutingOpts struct {
+	Addr       string
+	TreeHeight string
+}
+
+// TreeRoutingFactory defines the factory for tree routing
+type TreeRoutingFactory struct {
+	height      int
+	rootAddr    mino.Address
+	addrFactory mino.AddressFactory
+}
+
+// NewTreeRoutingFactory returns a new treeRoutingFactory. Warning: the given
+// rootAddr must be a uniq address that shouldn't be in the list of
+// participants.
+func NewTreeRoutingFactory(height int, rootAddr mino.Address,
+	addrFactory mino.AddressFactory) *TreeRoutingFactory {
+
+	return &TreeRoutingFactory{
+		height:      height,
+		rootAddr:    rootAddr,
+		addrFactory: addrFactory,
+	}
+}
+
+// FromIterator creates the network tree in a deterministic manner based on
+// the addresses.
+func (t TreeRoutingFactory) FromIterator(iterator mino.AddressIterator) (Routing, error) {
+
+	addrsBuf := make(addrsBuf, 0)
+	for iterator.HasNext() {
+		addr := iterator.GetNext()
+
+		addrBuf, err := addr.MarshalText()
+		if err != nil {
+			return nil, xerrors.Errorf("failed to marshal addr '%s': %v", addr, err)
+		}
+
+		addrsBuf = append(addrsBuf, addrBuf)
+	}
+
+	if len(addrsBuf) == 0 {
+		return nil, xerrors.Errorf("there should be at least one address in " +
+			"the iterator")
+	}
+
+	sort.Stable(&addrsBuf)
+
+	// We will use the hash of the addresses to set the random seed.
+	hash := sha256.New()
```

Comment:
the two functions have a lot of code repetition that could into a private function.

Reply:
done 👍